### ISSUE TITLE
Conversion to std::wstring produce invalid UTF-32 when wchar_t is 4 bytes

### ISSUE THREAD

**Issue by gracicot** [emscripten-core/emscripten#8954]
If any 4 byte characters are returned to javascript inside a `std::wstring`, the tests are failing:

Here's how I changed the tests:
```c++
std::wstring get_non_ascii_wstring() {
    std::wstring ws(5, 0);
    ws[0] = 10;
    ws[1] = 1234;
    ws[2] = 2345;
    ws[3] = 128513;
    ws[4] = 128640;
    return ws;
}
```
```js
test("non-ascii wstrings", function() {
    var expected = String.fromCharCode(10) +
        String.fromCharCode(1234) +
        String.fromCharCode(2345) +
        String.fromCharCode(55357) +
        String.fromCharCode(56833) +
        String.fromCharCode(55357) +
        String.fromCharCode(56960);
    assert.equal(expected, cm.get_non_ascii_wstring());
});
```

The test fails with this output:
>     string: non-ascii wstrings: FAIL
>     
>     Error: expected: 
>     Ӓऩ😁🚀, actual: 
>     Ӓऩ
>         at Object.assert.equal (/private/tmp/emscripten_test_other_ae7KFv/a.out.js:6777:18)
>         at Object.<anonymous> (/private/tmp/emscripten_test_other_ae7KFv/a.out.js:7619:20)
>         at runTest (/private/tmp/emscripten_test_other_ae7KFv/a.out.js:6610:23)
>         at run_all (/private/tmp/emscripten_test_other_ae7KFv/a.out.js:6630:26)
>         at run_all_tests (/private/tmp/emscripten_test_other_ae7KFv/a.out.js:7132:5)
>         at Object.<anonymous> (/private/tmp/emscripten_test_other_ae7KFv/a.out.js:9881:5)
>         at Module._compile (internal/modules/cjs/loader.js:738:30)
>         at Object.Module._extensions..js (internal/modules/cjs/loader.js:749:10)
>         at Module.load (internal/modules/cjs/loader.js:630:32)
>         at tryModuleLoad (internal/modules/cjs/loader.js:570:12)
>     error: { Error: expected: 
>     Ӓऩ😁🚀, actual: 
>     Ӓऩ
>         at Object.assert.equal (/private/tmp/emscripten_test_other_ae7KFv/a.out.js:6777:18)
>         at Object.<anonymous> (/private/tmp/emscripten_test_other_ae7KFv/a.out.js:7619:20)
>         at runTest (/private/tmp/emscripten_test_other_ae7KFv/a.out.js:6610:23)
>         at run_all (/private/tmp/emscripten_test_other_ae7KFv/a.out.js:6630:26)
>         at run_all_tests (/private/tmp/emscripten_test_other_ae7KFv/a.out.js:7132:5)
>         at Object.<anonymous> (/private/tmp/emscripten_test_other_ae7KFv/a.out.js:9881:5)
>         at Module._compile (internal/modules/cjs/loader.js:738:30)
>         at Object.Module._extensions..js (internal/modules/cjs/loader.js:749:10)
>         at Module.load (internal/modules/cjs/loader.js:630:32)
>         at tryModuleLoad (internal/modules/cjs/loader.js:570:12) info: { Expected: '\nӒऩ😁🚀', Actual: '\nӒऩ' } }

This is because valid UTF-32 `wchar_t` characters are translated with `String.fromCharCode` in `embind.js:
```js
var HEAP = getHeap();
var length = HEAPU32[value >> 2];
var a = new Array(length);
var start = (value + 4) >> shift;
for (var i = 0; i < length; ++i) {
    // The expression 'HEAP[start + i]' will be equal to 128513 when getting '😁'
    a[i] = String.fromCharCode(HEAP[start + i]);
}
```

This breaks the `String.fromCharCode` precondition:

> num1, ..., numN
> A sequence of numbers that are UTF-16 code units. The range is between 0 and 65535 (0xFFFF). Numbers greater than 0xFFFF are truncated. No validity checks are performed.

Taken from [`String.fromCharCode()` documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/fromCharCode)

So when passing 4 bytes `wchar_t` values, this code is out of bound.

Likewise, the same approach is taken for encoding:
```js
var HEAP = getHeap();
var length = value.length;
var ptr = _malloc(4 + length * charSize);
HEAPU32[ptr >> 2] = length;
var start = (ptr + 4) >> shift;
for (var i = 0; i < length; ++i) {
    HEAP[start + i] = value.charCodeAt(i);
}
if (destructors !== null) {
    destructors.push(_free, ptr);
}
return ptr;
```

This encode each UTF-16 into their own 4 byte character, including 2 character wide code point. This falsly makes make test passing because there is no 4 byte codepoints passed, but it results in invalid UTF-32, which is the assumed encoding for `wchar_t` on unix platforms.

**Comment by gracicot** [comment]
I tried changing the parsing of wstring to this function:
```js
_embind_register_std_wstring: function(rawType, charSize, name) {
  name = readLatin1String(name);
  var decodeString, encodeString;
  if (charSize === 2) {
      decodeString = UTF16ToString;
      encodeString = stringToUTF16;
  } else if (charSize === 4) {
      decodeString = UTF32ToString;
      encodeString = stringToUTF32;
  }
  registerType(rawType, {
      name: name,
      'fromWireType': function(value) {
          var start = (value + 4);
          var a = decodeString(start);
          _free(value);
          return a;
      },
      'toWireType': function(destructors, value) {
          var length = value.length;
          var ptr = _memalign(4, 4 + length * charSize);
          var start = (ptr + 4);
          HEAPU32[ptr >> 2] = encodeString(value, start, 4 + length * charSize) >> 4;
          if (destructors !== null) {
              destructors.push(_free, ptr);
          }
          return ptr;
       },
      'argPackAdvance': 8,
      'readValueFromPointer': simpleReadValueFromPointer,
      destructorFunction: function(ptr) { _free(ptr); },
  });
}
```

However the tests will output garbarge strings when decoding a wide string:

>     string: non-ascii wstrings: FAIL
>
>     Error: expected: 
>     Ӓऩ￿, actual: 
>     Ӓऩ￿��湯􋶠􋹀҈
>         at Object.assert.equal (/private/tmp/emscripten_test_other_qPOzyG/a.out.js:6735:18)
>         at Object.<anonymous> (/private/tmp/emscripten_test_other_qPOzyG/a.out.js:7574:20)
>         at runTest (/private/tmp/emscripten_test_other_qPOzyG/a.out.js:6568:23)
>         at run_all (/private/tmp/emscripten_test_other_qPOzyG/a.out.js:6588:26)
>         at run_all_tests (/private/tmp/emscripten_test_other_qPOzyG/a.out.js:7090:5)
>         at Object.<anonymous> (/private/tmp/emscripten_test_other_qPOzyG/a.out.js:9836:5)
>         at Module._compile (internal/modules/cjs/loader.js:738:30)
>         at Object.Module._extensions..js (internal/modules/cjs/loader.js:749:10)
>         at Module.load (internal/modules/cjs/loader.js:630:32)
>         at tryModuleLoad (internal/modules/cjs/loader.js:570:12)
>     error: { Error: expected: 
>     Ӓऩ￿, actual: 
>     Ӓऩ￿��湯􋶠􋹀҈
>         at Object.assert.equal (/private/tmp/emscripten_test_other_qPOzyG/a.out.js:6735:18)
>         at Object.<anonymous> (/private/tmp/emscripten_test_other_qPOzyG/a.out.js:7574:20)
>         at runTest (/private/tmp/emscripten_test_other_qPOzyG/a.out.js:6568:23)
>         at run_all (/private/tmp/emscripten_test_other_qPOzyG/a.out.js:6588:26)
>         at run_all_tests (/private/tmp/emscripten_test_other_qPOzyG/a.out.js:7090:5)
>         at Object.<anonymous> (/private/tmp/emscripten_test_other_qPOzyG/a.out.js:9836:5)
>         at Module._compile (internal/modules/cjs/loader.js:738:30)
>         at Object.Module._extensions..js (internal/modules/cjs/loader.js:749:10)
>         at Module.load (internal/modules/cjs/loader.js:630:32)
>         at tryModuleLoad (internal/modules/cjs/loader.js:570:12)
>       info:
>        { Expected: '\nӒऩ￿', Actual: '\nӒऩ￿\u0019��湯􋶠􋹀\u0018\u0012҈' } }

It seem to me like the strings are not null terminated when reading it. I did not investigated further yet.

**Comment by gracicot** [comment]
This issue has been closed by the pull request #9031
